### PR TITLE
Add override hook PGDialect.set_backslash_escapes()

### LINKED ISSUE #9442: we likely need to keep some mention of postgresql 8.2 around to have nominal redshift support
### Describe the bug

I believe there is a regression of https://github.com/sqlalchemy/sqlalchemy/issues/5698 in 2.0.0

I get the below error after updating from 1.4.46 and have confirmed it present in 2.0.0 - 2.0.5.post1
```
UndefinedObject: unrecognized configuration parameter "standard_conforming_strings"
```

The logic added in  https://github.com/sqlalchemy/sqlalchemy/commit/415ddce1e3b429bdc1180c5b5f7750a08d226525
does not appear to be present in 2.0.0

It appears to have been removed in https://github.com/sqlalchemy/sqlalchemy/commit/db08a699489c9b0259579d7ff7fd6bf3496ca3a2

as seen in 
Screenshot 2023-03-07 at 8 18 23 AM

### Optional link from https://docs.sqlalchemy.org which documents the behavior that is expected

_No response_

### SQLAlchemy Version in Use

2.0.5post1

### DBAPI (i.e. the database driver)

psycopg2

### Database Vendor and Major Version

Redshift 1.0.46987

### Python Version

3.10

### Operating system

OSX

### To Reproduce

```python
import sqlalchemy
from sqlalchemy.orm import sessionmaker
from sqlalchemy import text

connection_string = 'postgresql+psycopg2://<redacted username>:<redacted url encoded password>@<redacted db hostname>:<redacted port>/<redacted database>'

redshift_engine = sqlalchemy.create_engine(connection_string, echo=False, pool_recycle=120, pool_pre_ping=True)

redshift_session = sessionmaker(bind=redshift_engine)

tmp_session = redshift_session()
tmp_session.query(text("SELECT * from my_schema.my_table limit 10")).all()
```

### Error

```
Traceback (most recent call last):
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1964, in _exec_single_context
    self.dialect.do_execute(
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 747, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.UndefinedObject: unrecognized configuration parameter "standard_conforming_strings"
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/home/username/Downloads/pycharm-2022.3.2/plugins/python/helpers/pydev/pydevconsole.py", line 364, in runcode
    coro = func()
  File "<input>", line 1, in <module>
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2715, in all
    return self._iter().all()  # type: ignore
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2875, in _iter
    result: Union[ScalarResult[_T], Result[_T]] = self.session.execute(
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2229, in execute
    return self._execute_internal(
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2114, in _execute_internal
    conn = self._connection_for_bind(bind)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 1981, in _connection_for_bind
    return trans._connection_for_bind(engine, execution_options)
  File "<string>", line 2, in _connection_for_bind
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/orm/state_changes.py", line 137, in _go
    ret_value = fn(self, *arg, **kw)
  File "/home/username/.local/share/virtualenvs/Traceback (most recent call last):
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1964, in _exec_single_context
    self.dialect.do_execute(
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 747, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.UndefinedObject: unrecognized configuration parameter "standard_conforming_strings"
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/home/username/Downloads/pycharm-2022.3.2/plugins/python/helpers/pydev/pydevconsole.py", line 364, in runcode
    coro = func()
  File "<input>", line 1, in <module>
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2715, in all
    return self._iter().all()  # type: ignore
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2875, in _iter
    result: Union[ScalarResult[_T], Result[_T]] = self.session.execute(
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2229, in execute
    return self._execute_internal(
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2114, in _execute_internal
    conn = self._connection_for_bind(bind)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 1981, in _connection_for_bind
    return trans._connection_for_bind(engine, execution_options)
  File "<string>", line 2, in _connection_for_bind
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/orm/state_changes.py", line 137, in _go
    ret_value = fn(self, *arg, **kw)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 1108, in _connection_for_bind
    conn = bind.connect()
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 3245, in connect
    return self._connection_cls(self)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 145, in __init__
    self._dbapi_connection = engine.raw_connection()
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 3269, in raw_connection
    return self.pool.connect()
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 452, in connect
    return _ConnectionFairy._checkout(self)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 1255, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 716, in checkout
    rec = pool._do_get()
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/pool/impl.py", line 168, in _do_get
    with util.safe_reraise():
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py", line 147, in __exit__
    raise exc_value.with_traceback(exc_tb)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/pool/impl.py", line 166, in _do_get
    return self._create_connection()
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 393, in _create_connection
    return _ConnectionRecord(self)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 678, in __init__
    self.__connect()
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 916, in __connect
    )._exec_w_sync_on_first_run(self.dbapi_connection, self)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/event/attr.py", line 473, in _exec_w_sync_on_first_run
    self(*args, **kw)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/event/attr.py", line 487, in __call__
    fn(*args, **kw)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py", line 1912, in go
    return once_fn(*arg, **kw)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/engine/create.py", line 746, in first_connect
    dialect.initialize(c)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/dialects/postgresql/psycopg2.py", line 654, in initialize
    super().initialize(connection)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/dialects/postgresql/base.py", line 3028, in initialize
    std_string = connection.exec_driver_sql(
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1772, in exec_driver_sql
    ret = self._execute_context(
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1842, in _execute_context
    return self._exec_single_context(
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1983, in _exec_single_context
    self._handle_dbapi_exception(
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 2325, in _handle_dbapi_exception
    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1964, in _exec_single_context
    self.dialect.do_execute(
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 747, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.ProgrammingError: (psycopg2.errors.UndefinedObject) unrecognized configuration parameter "standard_conforming_strings"
[SQL: show standard_conforming_strings]
(Background on this error at: https://sqlalche.me/e/20/f405)-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 1108, in _connection_for_bind
    conn = bind.connect()
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 3245, in connect
    return self._connection_cls(self)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 145, in __init__
    self._dbapi_connection = engine.raw_connection()
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 3269, in raw_connection
    return self.pool.connect()
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 452, in connect
    return _ConnectionFairy._checkout(self)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 1255, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 716, in checkout
    rec = pool._do_get()
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/pool/impl.py", line 168, in _do_get
    with util.safe_reraise():
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py", line 147, in __exit__
    raise exc_value.with_traceback(exc_tb)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/pool/impl.py", line 166, in _do_get
    return self._create_connection()
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 393, in _create_connection
    return _ConnectionRecord(self)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 678, in __init__
    self.__connect()
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 916, in __connect
    )._exec_w_sync_on_first_run(self.dbapi_connection, self)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/event/attr.py", line 473, in _exec_w_sync_on_first_run
    self(*args, **kw)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/event/attr.py", line 487, in __call__
    fn(*args, **kw)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py", line 1912, in go
    return once_fn(*arg, **kw)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/engine/create.py", line 746, in first_connect
    dialect.initialize(c)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/dialects/postgresql/psycopg2.py", line 654, in initialize
    super().initialize(connection)
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/dialects/postgresql/base.py", line 3028, in initialize
    std_string = connection.exec_driver_sql(
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1772, in exec_driver_sql
    ret = self._execute_context(
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1842, in _execute_context
    return self._exec_single_context(
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1983, in _exec_single_context
    self._handle_dbapi_exception(
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 2325, in _handle_dbapi_exception
    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1964, in _exec_single_context
    self.dialect.do_execute(
  File "/home/username/.local/share/virtualenvs/project01-Z0F7Zsx7/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 747, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.ProgrammingError: (psycopg2.errors.UndefinedObject) unrecognized configuration parameter "standard_conforming_strings"
[SQL: show standard_conforming_strings]
(Background on this error at: https://sqlalche.me/e/20/f405)
``` 

### Additional context

Downgrading to 1.4.46 has provided a workaround

### PR DESCRIPTION
**Pull request by matthewgdv**
### Description
Refactor out the lines in `PGDialect.initialize()` that set backslash escapes into their own method to provide an override hook for [`sqlalchemy-redshift`](https://github.com/sqlalchemy-redshift/sqlalchemy-redshift) to use.

Fixes #9442

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [x] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

### CODE REVIEW

File: lib/sqlalchemy/dialects/postgresql/base.py [hunk 1]
```diff
@@ -3030,17 +3030,22 @@ def initialize(self, connection):
         # https://www.postgresql.org/docs/9.3/static/release-9-2.html#AEN116689
         self.supports_smallserial = self.server_version_info >= (9, 2)
 
-        std_string = connection.exec_driver_sql(
-            "show standard_conforming_strings"
-        ).scalar()
-        self._backslash_escapes = std_string == "off"
+        self.set_backslash_escapes()
```
Comment (zzzeek) [inline lib/sqlalchemy/dialects/postgresql/base.py:3033]: I think you need to pass the connection here

File: lib/sqlalchemy/dialects/postgresql/base.py [hunk 2]
```diff
@@ -3030,17 +3030,22 @@ def initialize(self, connection):
...
 
         self._supports_drop_index_concurrently = self.server_version_info >= (
             9,
             2,
         )
         self.supports_identity_columns = self.server_version_info >= (10,)
 
+    def set_backslash_escapes(self):
```
Comment (zzzeek) [inline lib/sqlalchemy/dialects/postgresql/base.py:3041]: this should be _set_backslash_escapes

**Comment by matthewgdv** [comment]
So, I'm not really sure what sort of tests to write for what is essentially just a refactor-as-method operation. Anything I can think of would probably be a duplicate of an existing test. Was hoping to maybe get some pointers as to what sort of test might be appropriate?

Thanks in advance.

**Review by zzzeek — CHANGES_REQUESTED** [review]
a few changes but this also needs a changelog file, of the form:

doc/build/changelog/unreleased_20/9442.rst

```
.. change::
      :tags: usecase, postgresql
      :tickets: 9442

      Modifications to the base PostgreSQL dialect to allow for better integration with the 
      sqlalchemy-redshift third party dialect for SQLAlchemy 2.0.  Pull request courtesy 
      matthewgdv.

```

File: lib/sqlalchemy/dialects/postgresql/base.py (hunk 2 above)
Comment (matthewgdv) [reply lib/sqlalchemy/dialects/postgresql/base.py:3041]: Should I also move it further down to the end of the class with the rest of the private methods? Unsure if there's a convention around method ordering.

File: lib/sqlalchemy/dialects/postgresql/base.py (hunk 1 above)
Comment (matthewgdv) [reply lib/sqlalchemy/dialects/postgresql/base.py:3033]: Derp. Thanks for catching that. My IDE wasn't done indexing sqlalchemy yet so the linter wasn't working yet xD

File: lib/sqlalchemy/dialects/postgresql/base.py (hunk 2 above)
Comment (zzzeek) [reply lib/sqlalchemy/dialects/postgresql/base.py]: not much, it's good enough

**Comment by zzzeek** [comment]
i wouldnt worry about a test here, our tests will pass, then on your end if you have a redshift patch, you need to test that one with this.

**Review by CaselIT — APPROVED** [review]

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 5565afeac20ea3612c3f427f58efacd8487ac159 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change 5565afeac20ea3612c3f427f58efacd8487ac159: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4508

**Comment by CaselIT** [comment]
Thanks the change looks good. Can you verify that it's enough for redshift to create a simple enough change?

@zzzeek not sure if it's feasible to backpork (or even when a 1.4 series release would be done)

**Comment by zzzeek** [comment]
IIUC redshift works with sqlalchemy 1.4 right now (somehow?)  whats the story there

**Comment by matthewgdv** [comment]
I believe it's because the logic in [this fix](https://github.com/sqlalchemy/sqlalchemy/commit/415ddce1e3b429bdc1180c5b5f7750a08d226525) was removed in the 2.0 release

**Comment by matthewgdv** [comment]
> Thanks, the change looks good. Can you verify that it's enough for redshift to create the needed customization?
> 
> @zzzeek not sure if it's feasible to backpork (or even when a 1.4 series release would be done)

I'm off work now but I'll look at the `sqlalchemy-redshift` side of things tomorrow. I'm about 99% sure that this should make the downstream fix trivial by implementing the new method as:

```
def _set_backslash_escapes(self, connection):
    self._backslash_escapes = False
```

Which is what the extra code path in the previous fix was doing

**Comment by matthewgdv** [comment]
So, I've made the change to `sqlalchemy-redshift` on my machine side-by-side with this change and can confirm I can open a connection and emit SQL statements against my redshift cluster again.

**Comment by CaselIT** [comment]
great, thanks for validating the change

**Comment by sqla-tester** [comment]
**Federico Caselli** (CaselIT) wrote:

should be ok

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4508

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4508 has been **merged**. Congratulations! :)

**Pull request closed without merging**